File: Day 6/solution.py
"""
Day 6 solution.
"""
import sys


class Solver:
    """Day 6 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def calculate_distance(self, total_time: int, time_held: int) -> int:
        """
        Calculate the total distance travelled when held for time_held and race is total_time long.
        """
        return time_held * (total_time - time_held)

    def find_minimum_and_maximum_valid_hold_times(
        self, race_time: int, minimum_distance: int
    ) -> tuple[int, int]:
        """
        Finds the maximum and minimum valid hold times.
        """
        # Find minimum valid hold time
        left, right = 0, race_time + 1
        while left < right:
            time = left + (right - left) // 2

            if self.calculate_distance(race_time, time) > minimum_distance:
                right = time
            else:
                left = time + 1
        minimum_hold_time = left

        # Find maximum valid hold time
        right = race_time
        while left < right:
            time = right - (right - left) // 2

            if self.calculate_distance(race_time, time) > minimum_distance:
                left = time
            else:
                right = time - 1
        maximum_hold_time = left

        return minimum_hold_time, maximum_hold_time

    def part_1(self):
        """Part 1 solver."""

        def parse_line(line: str) -> list[int]:
            return [int(x) for x in line.split(":")[1].split()]

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            race_times = parse_line(file.readline())
            distances = parse_line(file.readline())

        result = 1
        for race_time, distance in zip(race_times, distances):
            minimum, maximum = self.find_minimum_and_maximum_valid_hold_times(
                race_time, distance
            )
            result *= maximum - minimum + 1

        return result

    def part_2(self):
        """Part 2 solver."""

        def parse_line(line: str) -> int:
            return int("".join(line.split(":")[1].split()))

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            race_time = parse_line(file.readline())
            distance = parse_line(file.readline())

        minimum, maximum = self.find_minimum_and_maximum_valid_hold_times(
            race_time, distance
        )
        return maximum - minimum + 1

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 6")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case other:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()

File: Day 6/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>

class Solver
{
	std::filesystem::path filepath;

	long calculate_distance(long total_time, long time_held)
	{
		return time_held * (total_time - time_held);
	}

	std::pair<long, long> find_minimum_and_maximum_valid_hold_times(
			long race_time, long minimum_distance)
	{
		long left = 0, right = race_time + 1, time;

		// Find minimum valid hold time
		while (left < right)
		{
			time = left + (right - left) / 2;

			if (this->calculate_distance(race_time, time) > minimum_distance)
			{
				right = time;
			}
			else
			{
				left = time + 1;
			}
		}
		int minimum_hold_time = left;

		// Find maximum valid hold time
		right = race_time;
		while (left < right)
		{
			time = right - (right - left) / 2;

			if (this->calculate_distance(race_time, time) > minimum_distance)
			{
				left = time;
			}
			else
			{
				right = time - 1;
			}
		}
		int maximum_hold_time = left;

		return std::make_pair(minimum_hold_time, maximum_hold_time);
	}

	long part_1()
	{
		auto parse_line = [](std::ifstream &file) -> std::vector<long>
		{
			char c;

			// Skip labels
			while ((c = file.get()) != ':')
			{
			}
			while ((c = file.get()) == ' ')
			{
			}

			std::vector<long> result;
			long num = -1;
			for (; c != '\n' && c != EOF; c = file.get())
			{
				if (c == ' ')
				{
					if (num != -1)
						result.push_back(num);
					num = -1;
					continue;
				}

				if (num == -1)
					num = 0;
				num = num * 10 + c - '0';
			}
			result.push_back(num);

			return result;
		};

		std::ifstream file;
		file.open(this->filepath);
		std::vector<long> race_times = parse_line(file), distances = parse_line(file);
		file.close();

		long result = 1;
		for (int i = 0; i < race_times.size(); ++i)
		{
			auto [min, max] = this->find_minimum_and_maximum_valid_hold_times(race_times[i], distances[i]);
			result *= max - min + 1;
		}
		return result;
	}

	int part_2()
	{
		auto parse_line = [](std::ifstream &file) -> long
		{
			char c;

			// Skip labels
			while ((c = file.get()) != ':')
			{
			}
			while ((c = file.get()) == ' ')
			{
			}

			long result = 0;
			for (; c != '\n' && c != EOF; c = file.get())
			{
				if (c == ' ')
					continue;

				result = result * 10 + c - '0';
			}

			return result;
		};

		std::ifstream file;
		file.open(this->filepath);
		long race_time = parse_line(file), distance = parse_line(file);
		file.close();

		auto [min, max] = this->find_minimum_and_maximum_valid_hold_times(race_time, distance);
		return max - min + 1;
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 6" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}